File: src/GUI/objects/ImprovedScrollBarSkin.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package gui.objects;

import javafx.scene.control.ScrollBar;
import javafx.scene.layout.StackPane;

import com.sun.javafx.scene.control.skin.ScrollBarSkin;

import util.animation.Anim;

import static javafx.geometry.Orientation.VERTICAL;
import static javafx.scene.input.MouseEvent.MOUSE_ENTERED;
import static javafx.scene.input.MouseEvent.MOUSE_EXITED;
import static javafx.util.Duration.millis;
import static util.Util.getFieldValue;

/**
 * ScrollBar skin that adds animations & improved usability - thumb expands on mouse hover.
 * 
 * @author Plutonium_
 */
public class ImprovedScrollBarSkin extends ScrollBarSkin {

    public ImprovedScrollBarSkin(ScrollBar scrollbar) {
        super(scrollbar);

        StackPane thumb = getFieldValue(this, StackPane.class, "thumb");
        Anim v = new Anim(millis(350),p -> thumb.setScaleX(1+p*p));
        Anim h = new Anim(millis(350),p -> thumb.setScaleY(1+p*p));
        getSkinnable().addEventHandler(MOUSE_ENTERED, e -> {
            if(scrollbar.getOrientation()==VERTICAL) v.playOpen();
            else h.playOpen();
        });
        getSkinnable().addEventHandler(MOUSE_EXITED, e -> {
            if(scrollbar.getOrientation()==VERTICAL) v.playClose();
            else h.playClose();
        });
    }

}

File: src/gui/objects/ImprovedScrollBarSkin.kt
package gui.objects

import javafx.geometry.Orientation.VERTICAL
import javafx.scene.control.ScrollBar
import javafx.scene.control.skin.ScrollBarSkin
import javafx.scene.input.MouseEvent.DRAG_DETECTED
import javafx.scene.input.MouseEvent.MOUSE_ENTERED
import javafx.scene.input.MouseEvent.MOUSE_EXITED
import javafx.scene.input.MouseEvent.MOUSE_RELEASED
import javafx.scene.layout.StackPane
import javafx.util.Duration.millis
import util.animation.Anim
import util.type.Util.getFieldValue

/** ScrollBar skin that adds animations & improved usability - thumb expands on mouse hover. */
class ImprovedScrollBarSkin(scrollbar: ScrollBar): ScrollBarSkin(scrollbar) {
    private var isDragOn = false

    init {

        // install hover animation
        val thumb = getFieldValue<StackPane>(this, "thumb")!!
        val ah = Anim(millis(350.0)) { thumb.scaleY = 1+it*it }
        val av = Anim(millis(350.0)) { thumb.scaleX = 1+it*it }
        scrollbar.addEventHandler(MOUSE_ENTERED) { (if (scrollbar.orientation==VERTICAL) av else ah).playOpen() }
        scrollbar.addEventHandler(MOUSE_EXITED) {
            if (!isDragOn) {
                (if (scrollbar.orientation==VERTICAL) av else ah).playClose()
            }
        }
        scrollbar.addEventHandler(DRAG_DETECTED) { isDragOn = true }
        scrollbar.addEventHandler(MOUSE_RELEASED) {
            if (isDragOn) {
                isDragOn = false
                (if (scrollbar.orientation==VERTICAL) av else ah).playClose()
            }
        }
    }

}